x-pack/plugins/elastic_assistant/server/routes/post_actions_connector_execute.ts

```diff
@@ -8,139 +8,286 @@
 ) => {
-  router.post(
-    {
+  router.versioned
+    .post({
+      access: 'internal',
       path: POST_ACTIONS_CONNECTOR_EXECUTE,
-      validate: {
-        body: buildRouteValidation(PostActionsConnectorExecuteBody),
-        params: buildRouteValidation(PostActionsConnectorExecutePathParams),
+      options: {
+        tags: ['access:elasticAssistant'],
+      },
+    })
+    .addVersion(
+      {
+        version: ELASTIC_AI_ASSISTANT_INTERNAL_API_VERSION,
+        validate: {
+          request: {
+            body: buildRouteValidationWithZod(ExecuteConnectorRequestBody),
+            params: schema.object({
+              connectorId: schema.string(),
+            }),
+          },
+        },
       },
-    },
-    async (context, request, response) => {
-      const resp = buildResponse(response);
-      const assistantContext = await context.elasticAssistant;
-      const logger: Logger = assistantContext.logger;
-      const telemetry = assistantContext.telemetry;
-
-      try {
-        const connectorId = decodeURIComponent(request.params.connectorId);
-
-        // get the actions plugin start contract from the request context:
-        const actions = (await context.elasticAssistant).actions;
-
-        // if not langchain, call execute action directly and return the response:
-        if (
-          !request.body.isEnabledKnowledgeBase &&
-          !requestHasRequiredAnonymizationParams(request)
-        ) {
-          logger.debug('Executing via actions framework directly');
-          const result = await executeAction({ actions, request, connectorId });
+      async (context, request, response) => {
+        const resp = buildResponse(response);
+        const assistantContext = await context.elasticAssistant;
+        const logger: Logger = assistantContext.logger;
+        const telemetry = assistantContext.telemetry;
+
+        try {
+          const authenticatedUser = assistantContext.getCurrentUser();
+          if (authenticatedUser == null) {
+            return response.unauthorized({
+              body: `Authenticated user not found`,
+            });
+          }
+          const dataClient = await assistantContext.getAIAssistantConversationsDataClient();
+
+          let onMessageSent;
+          let prevMessages;
+          const conversationId = request.body.conversationId;
+          if (conversationId) {
+            const conversation = await dataClient?.getConversation({
+              id: conversationId,
+              authenticatedUser,
+            });
+            prevMessages = conversation?.messages?.map((c) => ({
+              role: c.role,
+              content: c.content,
+            }));
+
+            if (conversation == null) {
+              return response.notFound({
+                body: `conversation id: "${conversationId}" not found`,
+              });
+            }
+
+            if (request.body.replacements) {
+              await dataClient?.updateConversation({
```

Comment:
I'm wondering if it's necessary to update the conversation with replacements from the client here, before the server has generated a response.

It is _very_ important to merge replacements from the client, which may contain new values that the server doesn't know about, because (for example) the user may have added an alert from the client side, with results from the server, and persist those merged results. But that appears to happen later in the code.